File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;
vector<int> solution(vector<string> operations) {
    vector<int> answer;
    deque<int> dq;
    
    for(int i=0; i<operations.size();i++){
        if(operations[i][0] == 'I'){    // insert
            int input = stoi(operations[i].substr(2));
            dq.push_back(input);
        }
        else if(operations[i][0] == 'D'){ 
            if(dq.empty()) continue;
            else if(dq.size() > 1) sort(dq.begin(), dq.end());

            if(operations[i][2] == '1'){    // max
                dq.pop_back();
            }else{  // min
                dq.pop_front();
            }
        }
    }
    if(dq.empty()){
        answer.push_back(0);
        answer.push_back(0);
    }
    else{
        if(dq.size() > 1) sort(dq.begin(), dq.end());
        answer.push_back(dq.back());
        answer.push_back(dq.front()); 
    }

    return answer;
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
    public int[] solution(string[] operations) {
        int[] answer = new int[] {0,0};
        List<int> arr = new List<int>();
        
        foreach(string operation in operations){
            if(operation[0] == 'I'){
                int num = int.Parse(operation.Substring(2));
                arr.Add(num);
                arr.Sort();
            }
            else{
                if(arr.Count == 0) continue;
                else if(int.Parse(operation.Substring(2)) == 1){
                    arr.Remove(arr.Max());
                }else{
                    arr.Remove(arr.Min());
                }
            }
        }
        
        if(arr.Count > 0){
            answer[0] = arr.Max();
            answer[1] = arr.Min();
        }
        return answer;
    }
}